File: observation-client/src/reducer.ts
import { EventTypes, type EvolutionEventEnvelope } from './contracts/envelope.js';
export interface ObservationUiState {
  facets: { isr?: unknown; fitness?: unknown; candidates?: unknown[]; governance?: unknown[]; evolution?: unknown; };
  meta: { generation: number; facetUpdatedAt: Record<string, string> };
}
export function initialUiState(): ObservationUiState {
  return { facets: {}, meta: { generation: 0, facetUpdatedAt: {} } };
}
const FACET_FOR_EVENT: Record<string, string> = {
  [EventTypes.IsrUpdated]: 'isr',
  [EventTypes.FitnessEvaluated]: 'fitness',
  [EventTypes.CandidatePromoted]: 'candidates',
  [EventTypes.GovernanceDecisionMade]: 'governance',
  [EventTypes.EvolutionStageChanged]: 'evolution',
};
export function observationReducer(state: ObservationUiState, envelope: EvolutionEventEnvelope): ObservationUiState {
  const facet = FACET_FOR_EVENT[envelope.eventType];
  if (!facet) return state;
  const facets = applyFacet(state.facets, facet, envelope.payload);
  return { facets, meta: { generation: envelope.generation, facetUpdatedAt: { ...state.meta.facetUpdatedAt, [facet]: envelope.occurredAt } } };
}
function applyFacet(facets: ObservationUiState['facets'], facet: string, payload: unknown): ObservationUiState['facets'] {
  if (facet === 'candidates') {
    const list = [...(facets.candidates ?? [])];
    const cid = (payload as { candidateId?: string } | null)?.candidateId;
    const idx = cid != null ? list.findIndex((c) => (c as { candidateId?: string })?.candidateId === cid) : -1;
    if (idx >= 0) list[idx] = payload; else list.push(payload);
    return { ...facets, candidates: list };
  }
  if (facet === 'governance') return { ...facets, governance: [...(facets.governance ?? []), payload] };
  return { ...facets, [facet]: payload };
}

File: autonomous-api/app/observation/readmodel/reducer.py
"""CompositeObservationReducer -- platform fold, heartbeat no-op."""
from __future__ import annotations

def initial_state():
    return {"facets": {}, "meta": {"generation": 0, "facetUpdatedAt": {}}}

FACET_FOR_EVENT = {
    "fitness.evaluated": "fitness",
    "candidate.promoted": "candidates",
    "governance.decision_made": "governance",
    "isr.updated": "isr",
    "evolution.stage_changed": "evolution",
}

class CompositeObservationReducer:
    def fold(self, state, envelope):
        facet = FACET_FOR_EVENT.get(envelope.eventType)
        if not facet:
            return state
        facets = dict(state["facets"])
        if facet == "candidates":
            lst = list(facets.get("candidates", []))
            cid = getattr(envelope.payload, "get", lambda k, d=None: None)("candidateId") if isinstance(envelope.payload, dict) else getattr(envelope.payload, "candidateId", None)
            idx = next((i for i, c in enumerate(lst) if isinstance(c, dict) and c.get("candidateId") == cid), -1)
            if idx >= 0:
                lst[idx] = envelope.payload
            else:
                lst.append(envelope.payload)
            facets["candidates"] = lst
        elif facet == "governance":
            facets["governance"] = list(facets.get("governance", [])) + [envelope.payload]
        else:
            facets[facet] = envelope.payload
        meta = dict(state["meta"])
        meta["generation"] = envelope.generation
        meta["facetUpdatedAt"] = dict(meta.get("facetUpdatedAt", {}))
        meta["facetUpdatedAt"][facet] = envelope.occurredAt
        return {"facets": facets, "meta": meta}
